src/main/java/htsjdk/samtools/SAMRecordDuplicateComparator.java

```diff
@@ -57,27 +60,44 @@
     
     public SAMRecordDuplicateComparator() {}
 
+    @Deprecated // not thread safe. will be removed in future version
     public SAMRecordDuplicateComparator(final List<SAMFileHeader> headers) {
-        // pre-populate the library names
-        for (final SAMFileHeader header : headers) {
-            for (final SAMReadGroupRecord readGroup : header.getReadGroups()) {
-                final String libraryName = readGroup.getLibrary();
-                if (null != libraryName) {
-                    final short libraryId = this.nextLibraryId++;
-                    this.libraryIds.put(libraryName, libraryId);
-                }
+        for (SAMFileHeader header : headers) {
```

Comment:
final

Reply:
I'm worried now about the interaction between the UNKNOWN_LIBRARY and this for loop, since at each iteration the UNKNOWN_LIBRARY will be added to the libraryIds Map. see idea below.

Reply:
I'm not sure I entirely understand what you are looking for here.  The libraryNameOrder is a set, so if it gets added multiple times, it doesn't have any impact on the set.